api/src/main/java/org/apache/iceberg/expressions/ResidualEvaluator.java

```diff
@@ -195,6 +196,26 @@ public Expression or(Expression leftResult, Expression rightResult) {
       return (cmp.compare(ref.eval(struct), lit.value()) != 0) ? alwaysTrue() : alwaysFalse();
     }
 
+    @Override
+    public <T> Expression in(BoundReference<T> ref, Set<T> literalSet) {
+      T val = ref.eval(struct);
+      if (literalSet.stream().anyMatch(lit -> ref.comparator().compare(val, lit) == 0)) {
```

Comment:
The literal set can handle all types, so this can use `contains` instead of checking each value individually.